spray-json/shared/src/main/scala/spray/json/JsonParser.scala

```diff
@@ -165,51 +170,148 @@ class JsonParser(input: ParserInput, settings: JsonParserSettings = JsonParserSe
   private def DIGIT(): Boolean = cursorChar >= '0' && cursorChar <= '9' && advance()
 
   // http://tools.ietf.org/html/rfc4627#section-2.5
-  private def `string`(): Unit = {
-    if (cursorChar == '"') cursorChar = input.nextUtf8Char() else fail("'\"'")
-    sb.setLength(0)
-    while (`char`()) cursorChar = input.nextUtf8Char()
-    require('"')
+  private def `string`(): String = {
+    if (cursorChar != '"') fail("'\"'")
+    val res = stringFast(input.cursor)
     ws()
+    res
   }
 
-  private def `char`() =
-    // simple bloom-filter that quick-matches the most frequent case of characters that are ok to append
-    // (it doesn't match control chars, EOI, '"', '?', '\', 'b' and certain higher, non-ASCII chars)
-    if (((1L << cursorChar) & ((31 - cursorChar) >> 31) & 0x7ffffffbefffffffL) != 0L) appendSB(cursorChar)
-    else cursorChar match {
-      case '"' | EOI => false
-      case '\\' =>
-        advance(); `escaped`()
-      case c => (c >= ' ') && appendSB(c)
+  var charBuffer: Array[Char] = JsonParser.charBufferCache.get()
+
+  private def stringFast(start: Int): String = {
+    /* Scan to the end of a simple string as fast as possible, no escapes, no unicode */
+    // so far unclear results if that is really helpful
+    /*@tailrec def parseOneFast(cursor: Int, charsRead: Int): String = {
+      val b = input.byteAt(cursor)
```

Comment:
I had another try and found out that `parseOneFast` was slower because I wasn't using `charAt` here directly as below in `parseOneSlow`. Changing that it now makes sense again to keep it to improve parsing speed of simple strings (without escapes and unicode).

Comparing to jsoniter-scala I found another weird thing which is that surrounding the whole `parseOneFast` with a useless (always true) `if` clause (`if (i < minLim)` in jsoniter and here e.g. `if(charsRead <= cursor)`) is another 20-25% improvement.

Those kind of things seem to be rather random findings that cannot really be explained in terms of the original code but only by looking at all stages of the compilation/execution pipeline (Scala compiler -> JIT compiler -> µops). I tried Intel's VTune for the first time to get some insights and while it looks promising on the surface I didn't find any actionable explanations for the performance cliffs (and in any case it those could be quite specific to my particular processor). E.g. for the `charAt` vs. `byteAt` it showed bad core port usage ratios (>50% of cycles where no port made progress) which (AFAIU) may be caused by too many dependencies between instructions.

(So, kudos again to jsonitor-scala for managing to be fast in various environments but as I said before that's not my goal here)

Reply:
crazy stuff... and a huge time sink, unfortunately